internal/daemon/api_logs.go

```diff
@@ -0,0 +1,307 @@
+	var fifo chan servicelog.Entry
+	if numLogs > 0 {
+		fifo = make(chan servicelog.Entry, numLogs)
+	}
+	flushFifo := func() error { // helper to flush any logs in the FIFO
+		if numLogs <= 0 || len(fifo) == 0 {
+			return nil
+		}
+		var err error
+		for len(fifo) > 0 && err == nil {
+			err = encoder.Encode(jsonLog(<-fifo))
+		}
+		if err != nil {
+			logger.Noticef("error writing logs: %v", err)
+			return err
+		}
+		flushWriter(w)
+		return nil
+	}
+
+	// Background goroutine to stream ordered logs: it sends parsed logs on
+	// logs channel, any error on errorChan channel, and stops when the request
+	// is cancelled or this handler returns.
+	logs := make(chan servicelog.Entry)
+	errorChan := make(chan error, 1)
+	ctx, cancel := context.WithCancel(req.Context())
+	defer cancel()
+	go func() {
+		errorChan <- streamLogs(itsByName, logs, ctx.Done())
+	}()
+
+	// Main loop: output earliest log per iteration. Stop when request
+	// cancelled or there are no more logs (in non-follow mode).
+	requestStarted := time.Now().UTC()
+	for {
+		select {
+		case log := <-logs:
+			if log.Time.IsZero() {
+				// Zero-time log means we've consumed all buffered logs
+				if flushFifo() != nil {
+					return
+				}
+				if follow {
+					// Following, wait for more
+					numLogs = 0 // so we don't use the FIFO from here on
+					continue
+				}
+				// Not following, we're done
+				return
+			}
+
+			// Logs are coming faster than we can send them (probably a slow
+			// client), so stop now.
+			if !follow && log.Time.After(requestStarted) {
+				_ = flushFifo()
+				return
+			}
+
+			if numLogs > 0 {
+				// Push through FIFO so we only output the first "n" across
+				// all services.
+				if len(fifo) == cap(fifo) {
+					// FIFO channel full, discard oldest log entry before
+					// writing new one so it doesn't block.
+					<-fifo
+				}
+				fifo <- log
+				continue
+			}
+
+			// Otherwise encode and output log directly.
+			err := encoder.Encode(jsonLog(log))
+			if err != nil {
+				logger.Noticef("error writing logs: %v", err)
+				return
+			}
+			if follow {
+				flushWriter(w)
+			}
+
+		case err := <-errorChan:
+			logger.Noticef("%s", err)
+			return
+
+		case <-req.Context().Done():
+			return
+		}
+	}
+}
+
+// streamLogs reads and parses logs from the given services, merging the
+// log streams and ordering by timestamp. It sends the parsed logs to the
+// logs channel, and returns when the done channel is closed.
+func streamLogs(itsByName map[string]servicelog.Iterator, logs chan<- servicelog.Entry, done <-chan struct{}) error {
+	// Need to close iterators in same goroutine we're reading them from.
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Make a channel and register it with each of the iterators to be
+	// notified when new data comes in. We don't strictly need this when not
+	// following, but it doesn't hurt either.
+	notification := make(chan bool, 1)
```

Comment:
The notification here is to "wake up" the main processing loop -- it's only for knowing if there's more data on any one of the iterators (and we check them all). The producer here is `RingBuffer.signalIterators`, but the send to `notifyChan` doesn't block (it has a default case). So if we had no buffer it'd miss notifications, but a buffer size of 1 is all we need. `Iterator.Notify` is currently commented as follows:

```go
	// Notify sets the notification channel. When more data is available, the channel
	// passed in to Notify will have true sent on it. If the channel is not receiving (unbuffered)
	// or full (buffered), the notification will be dropped.
```

Because the buffer size is 1 some notifications will be dropped, but that's okay, the main loop will at least be notified once and pick up any more when it checks the iterators. Hope that helps.